### DISCUSSION TITLE
How to route all keyboard events to a single widget?

### DISCUSSION THREAD

**Discussion by dkogan** [fltk/fltk#1048] [Q&A] [answered]
Hi.

Hi. I'm trying to build an fltk application where all keyboard interactions do the same thing, regardless of which widget is focused. I'm having some trouble, and would appreciate being pointed in the right direction. A sample program appears below. This has 5 buttons and a slider. If a user presses FL_Right I'd like the slider to advance, whether it's focused or not. Here I'm explicitly asking the buttons to ignore Fl_Right by returning 0 it their handle(). But this doesn't work: the navigation code interprets Fl_Right, and the application focuses the next widget in response.

The sample program has the slider print out all keyboard events that do reach it, and we see that keys other than those that can be interpreted as navigation keys DO reach it properly. Thanks!

I'm on Debian, using fltk 1.3. Building and running the sample like this: `g++ -o fltk-buttons-slider fltk-buttons-slider.cc -lfltk && ./fltk-buttons-slider`

```c
#include <FL/Fl_Button.H>
#include <FL/Fl_File_Chooser.H>
#include <FL/Fl_Value_Slider.H>

class Fl_Button_noright : public Fl_Button
{
public:
    Fl_Button_noright(int x, int y, int w, int h, const char* label)
        : Fl_Button(x,y,w,h,label)
    {
    }

    int handle(int event)
    {
        if((event == FL_KEYDOWN || event == FL_SHORTCUT) &&
           Fl::event_key() == FL_Right)
        {
            printf("Button trying to ignore FL_KEYDOWN Fl_Right\n");
            return 0;
        }
        return Fl_Button::handle(event);
    }
};

class Fl_Value_Slider_report_keys : public Fl_Value_Slider
{
public:
    Fl_Value_Slider_report_keys(int x, int y, int w, int h, const char* label)
        : Fl_Value_Slider(x,y,w,h,label)
    {
    }

    int handle(int event)
    {
        if(event == FL_KEYDOWN)
            printf("slider got FL_KEYDOWN '%04x'\n", Fl::event_key());
        if(event == FL_SHORTCUT)
            printf("slider got FL_SHORTCUT '%04x'\n", Fl::event_key());
        return Fl_Value_Slider::handle(event);
    }
};

int main(int argc, char* argv[])
{
    Fl_Window window (800, 600, "window");

    int                         iy = 0;
    Fl_Button_noright           button1(0, (iy++)*100, 800,100, "button1");
    Fl_Button_noright           button2(0, (iy++)*100, 800,100, "button2");
    Fl_Button_noright           button3(0, (iy++)*100, 800,100, "button3");
    Fl_Button_noright           button4(0, (iy++)*100, 800,100, "button4");
    Fl_Button_noright           button5(0, (iy++)*100, 800,100, "button5");
    Fl_Value_Slider_report_keys slider( 0, (iy++)*100, 800,100, "slider");
    slider.type(FL_HOR_SLIDER);

    window.resizable(window);
    window.end();
    window.show();

    Fl::run();
    return 0;
}
```

**Comment by Albrecht-S** [comment, marked as answer]
1. Here's a quick solution if you want this effect only for a specific application: set `visible_focus()` off (0) for class `Fl_Button_noright` in its constructor:
```C++
    Fl_Button_noright(int x, int y, int w, int h, const char* label)
        : Fl_Button(x,y,w,h,label)
    {
        visible_focus(0);
    }
```

Tested with FLTK 1.4 and works but FLTK 1.3 shouldn't make a difference.

You can achieve the same for normal `Fl_Button`'s by setting `buttonX->visible_focus(0)` explicitly. This inhibits navigation by arrow keys. AFAICT Fl_Button's ignore arrow keys anyway.

2. A more involved but still "pretty simple" (TM) solution is to derive a class from `Fl_Group` and "insert" it between the window and the buttons, i.e. add this group to your window and add all buttons and the slider to that group. This group's `handle()` method would catch the arrow keys and delegate these events to one known child (the slider widget), something like (untested):
```C++
MyGroup::handle(int event) {
  if ((event == FL_KEYDOWN || event == FL_SHORTCUT || event == FL_KEYUP) &&
      (Fl::event_key == FL_Left || Fl::event_key == FL_Right)) {
    // optional: give the slider the focus (commented out):
    // slider->take_focus();
    slider->handle(event); // slider is a well-known child of the group
    return 1;
  }
  return Fl_Group::handle(event);
}
```

Note 1: This `handle()` method intentionally eclipses the base class `Fl_Group` from arrow key events.
Note 2: Obviously you could also do this in a subclass of Fl_(Double_)Window w/o the intermediate group.

I'm leaving the details to the reader...

**Reply by dkogan** [reply]
I tried it out, and both methods work very well. Thank you very much! A bit of detail for those reading later:

- Using `visible_focus(0)` in the buttons makes it such the buttons never appear focused, and the only focusable widget remaining (the slider) gets the visible focus and never loses it
- Using a custom group or window to explicitly route the messages the way we want still allows the buttons to get focus
